cpp/src/arrow/io/caching.cc

```diff
@@ -193,6 +195,36 @@ Result<std::shared_ptr<Buffer>> ReadRangeCache::Read(ReadRange range) {
   return Status::Invalid("ReadRangeCache did not find matching cache entry");
 }
 
+Future<> ReadRangeCache::Wait() {
+  struct State {
+    explicit State(std::vector<Future<std::shared_ptr<Buffer>>> f)
+        : futures(std::move(f)), remaining(futures.size()) {}
+    std::vector<Future<std::shared_ptr<Buffer>>> futures;
+    std::atomic<size_t> remaining;
+  };
+
+  std::vector<Future<std::shared_ptr<Buffer>>> futures;
+  for (const auto& entry : impl_->entries) {
+    futures.push_back(entry.future);
+  }
+  auto out = Future<>::Make();
+  auto state = std::make_shared<State>(std::move(futures));
+  for (const auto& future : state->futures) {
+    future.AddCallback([state, out](const Result<std::shared_ptr<Buffer>>&) mutable {
+      if (state->remaining.fetch_sub(1) != 1) return;
```

Comment:
Ideally this primitive would be globally available, instead of reimplemented here. So you could write for example:
```c++
  return Future::Gather(std::move(futures));
```

Also there may be different kinds of "gather". By default you probably want to first error to immediately mark the resulting future errored.

cc @westonpace

Reply:
Agreed. I can move this to future.h/future.cc since there's already a very similar helper there.

Reply:
What is the difference between this and `All` inside `future.h`?

Reply:
All gives `Future<vector<Result<T>>>` and never marks the future failed; this gives `Future<>` and marks the future failed if any individual result failed. Additionally this 'fails fast' in that if any fail, you immediately get a result, instead of waiting for the rest of the futures.

Reply:
I wonder if it might be better to add `fail_fast` as an optional flag to the existing method.  We may someday want `Future<vector<Result<T>>>` but also want the `fail_fast` behavior.

Also, just speculation, but I wonder if it would be possible to have a cast (not sure if explicit or implicit) from Future<Any> -> Future<>.  Under the hood it could be `base.Then([] (const Result<T>& result) -> Status {return Status::OK();})` (or perhaps there is a more efficient way to achieve the same thing).

Then it could just be `static_cast<Future<>>(All(futures, Future::FailFast))`

Reply:
Future<T> is already convertible to Future<>. So what you suggest would work, but we'd need the fail-fast behavior, yes. 

A flag like that, though, does change the semantics rather dramatically, I'd rather keep them as separate functions.